### ISSUE TITLE
HTML/CSS/JS Fedback

### ISSUE THREAD

**Issue by rkojik** [Office-of-Digital-Innovation/California-State-Web-Template#10]
Would like to start out by saying this is a good start! I think we're going in a great direction with "one look" to help the people of California who depend on our services. That in mind, technology like the template should help us work with agility to improve our website / services. I have a couple of concerns: 

- Markup, styles, and script (to a lesser extent) generally perform very well, but with an increase in mobile users with less bandwidth, it's important to not only keep the file size low, but reduce the amount of calls to the server. 

- As you're probably aware, file size isn't just important for performance, but implementation as well. When we're building pages, the more we have to navigate through and learn, the longer it takes to build. I know we want to limit "custom styles", but overriding styles with more styles isn't a best practice, adds unnecessary complexity, and increases maintenance cost.

- Keeping things simple isn't just a performance concern, it's also an accessibility concern. Basic HTML is extremely accessible and usable. The more we deviate from simple HTML, the higher chance we have of breaking it. 

Here are some examples: 
**A combined 16,000 lines of CSS is too much.** 
An audit (Chrome Dev tools) of a sample template page: 2932 rules (88%) of CSS is not used by the current page. 

I believe almost all of this can be alleviated by getting rid of Bootstrap. I understand it's a very popular framework, but it has a lot of bloat. What's worse, it requires over opinionated markup. Please, please, please consider using a leaner framework. 

- 18F advises [Bourbon](http://bourbon.io/) and [Neat](http://neat.bourbon.io/) (grid framework). 

- I HIGHLY recommend taking a look at Heydon Pickering's work. Heydon is doing phenomenal stuff in the industry. Not just making sure user's can access stuff, but also making sure it's useful: 
[Practical ARIA examples](http://heydonworks.com/practical_aria_examples/)
[Inclusive Components](https://inclusive-components.design/)
Please pardon the not safe for government naming convention, but this grid is the best I've seen. It's so simple it fits in a tweet!: 
[Grid]( https://github.com/Heydon/fukol-grids)

**Complicated selectors**
A "CSS Dig" yields several overly specific selectors, for example: 
`.header-slideshow-banner #primary-carousel .slide-text .title{
        font-size: 19px;
    }`

That is 140 specificity to change a font-size. My personal preference is to use relative font-sizing on everything. Then use one media query to change the font-size percentage of the HTML. This will increase or decrease the font-size relatively for everything, and reduce a lot of unnecessary font size declarations. Example: 
```
html {
font-size: 62.5%; 
}

body {
font-size: 2.0em; /*20 px*/ 
}

@media (max-width: 767px) {
html {
font-size: 56.3%; 
}
}
```

The best way I can describe this selector, is "WHAT??!?!?!?!" (laugh out loud): 
```
.panel > .table:first-child > thead:first-child > tr:first-child td:first-child
.panel > .table-responsive:first-child > .table:first-child > thead:first-child > tr:first-child td:first-child
.panel > .table:first-child > tbody:first-child > tr:first-child td:first-child
.panel > .table-responsive:first-child > .table:first-child > tbody:first-child > tr:first-child td:first-child
.panel > .table:first-child > thead:first-child > tr:first-child th:first-child
.panel > .table-responsive:first-child > .table:first-child > thead:first-child > tr:first-child th:first-child
.panel > .table:first-child > tbody:first-child > tr:first-child th:first-child
.panel > .table-responsive:first-child > .table:first-child > tbody:first-child > tr:first-child th:first-child{
    border-top-left-radius: -1;
}
```
That's very complicated. It might be an odd question, but why would we want to: 

> select any td element that is a first child that is a descendant of a tr element that is a first child that is a child of a thead element that is a first child that is a child of any element with a class attribute that contains the word table and that is a first child that is a child of any element with a class attribute that contains the word panel?

I'm not even 100% sure it does anything? No unit of measurement declared, and negative value border radius? And there are almost 200 more selectors with 50+ specificity. 

For simplicity, I generally try to stay in the 20-30 range at MOST. 

**Overuse of classes in markup**
```
<li class="nav-item active">
<a href="/sample/" class="first-level-link has-sub">...</a>
</li>
```

A class "has-sub" does nothing to help people with assistive technology. And "first-level-link" isn't necessary. We can tell what a first level link is by something like .navigation > li, and use useful markup to inform user's with assistive technology that there is a sub menu: 

```
<a href="/sample/" aria-haspopup="true">
<li aria-hidden="true" aria-label="submenu">...</li>
</a>
```
Then you can style like so: 
`[aria-haspopup] {
padding: 1em;
}
`

It's very important keep our markup and styles as simple / clean as possible! 

**A Lighthouse scan:** 

> Link elements are blocking the first paint of your page. Consider inlining critical links and deferring non-critical ones. [Learn more](https://developers.google.com/web/tools/lighthouse/audits/blocking-resources).

Basically we need to load CSS before script, and reduce the amount of CSS files loaded. 

---------------------------------------------------------------------------------------------------

I hope this helps, not hurts... I'm all-in with you in trying to make the template for CA great for our users! Please feel free to call or email me if you have any questions, comments, or concerns: 

Koji Kumpulainien
Franchise Tax Board 
916-845-4236 
koji.kumpulainien@ftb.ca.gov

**Comment by artuoma** [comment]
Great stuff, thanks Koji, we will be in touch!

**Comment by Brian-CalVCB** [comment]
> …it's important to not only keep the file size low, but reduce the amount of calls to the server.

Agreed. For example, we could save one request by rolling the color scheme into `cagov.core.css` during the build process rather than downloading it separately. (I'll probably make a pull request for this soon.)

> The best way I can describe this selector, is "WHAT??!?!?!?!" (laugh out loud):
> 
> .panel > .table:first-child > thead:first-child > tr:first-child td:first-child
> .panel > .table-responsive:first-child > .table:first-child > thead:first-child > tr:first-child td:first-child
> .panel > .table:first-child > tbody:first-child > tr:first-child td:first-child
> .panel > .table-responsive:first-child > .table:first-child > tbody:first-child > tr:first-child td:first-child
> .panel > .table:first-child > thead:first-child > tr:first-child th:first-child
> .panel > .table-responsive:first-child > .table:first-child > thead:first-child > tr:first-child th:first-child
> .panel > .table:first-child > tbody:first-child > tr:first-child th:first-child
> .panel > .table-responsive:first-child > .table:first-child > tbody:first-child > tr:first-child th:first-child{
>     border-top-left-radius: -1;
> }

Funny enough, this is straight out of Bootstrap, and it actually makes sense. It's meant to target the corners of a table that's in a certain position in a panel.

If a cell is the first child of many first children, it will be the top-left corner. If it's the last child of many first children, it will be the top-right corner, and so on.

But yeah, that selector still looks goofy.

The unexpected `-1` value is because the `@panel-border-radius` variable was set to `0` in `variables.less` rather than the Bootstrap default, which seems to be `4px`.

> A combined 16,000 lines of CSS is too much…. I believe almost all of this can be alleviated by getting rid of Bootstrap.

Agreed. I'm thinking a lot of the CSS bloat can be managed simply by commenting out modules that your site won't use in `cagov.core.less`. When you run the build process again, `cagov.core.css` will be a lot smaller.

**Comment by rkojik** [comment]
> If a cell is the first child of many first children, it will be the top-left corner. If it's the last child of many first children, it will be the top-right corner, and so on.

Thanks for the breakdown Brian! Awesome. :+1. 

My goal was more to point out Bootstrap's gross over specification and lack of understanding of the 'C' (cascade) part of CSS. 

**If our design is simple, it's much simpler to code**. That exact selector could have been: `.panel tr:first-child td:first-child`. But even that is too specific. I feel the need to call-out the fact that they defined a "tr" class for a table row... (painfully bad it deserves infinite eye rolls - haha) And they do that over and over and over. 

 If we want a border radius on a table, we use the table: `table {border-radius: 4px;}`, And because we used an element selector, if for some reason we have "special styled" tables, we should first ask ourselves why we need a special style in the first place): then you can easily override it with a useful attribute (no need for a class): 
`[aria-describedby^="table-"]`. 

Then, have a `<p id="table-pricing-description">` with useful information about that table. That is just one instance where we can make our markup more accessible, then use it as a simpler style hook.

**Comment by Brian-CalVCB** [comment]
> we could save one request by rolling the color scheme into cagov.core.css during the build process rather than downloading it separately. (I'll probably make a pull request for this soon.)

Here's that pull request: #17.
